hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -97,31 +104,170 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
   virtual CallbackReturn on_init(const HardwareInfo & hardware_info)
   {
     info_ = hardware_info;
+    import_state_interface_descriptions(info_);
+    import_command_interface_descriptions(info_);
+    create_report_interfaces();
     return CallbackReturn::SUCCESS;
   };
 
+  /**
+   * Import the InterfaceDescription for the StateInterfaces from the HardwareInfo.
+   * Separate them into the possible types: Joint and store them.
+   */
+  virtual void import_state_interface_descriptions(const HardwareInfo & hardware_info)
+  {
+    auto joint_state_interface_descriptions =
+      parse_joint_state_interface_descriptions_from_hardware_info(hardware_info);
+    for (const auto & description : joint_state_interface_descriptions)
+    {
+      joint_state_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
+  /**
+   * Import the InterfaceDescription for the CommandInterfaces from the HardwareInfo.
+   * Separate them into the possible types: Joint and store them.
+   */
+  virtual void import_command_interface_descriptions(const HardwareInfo & hardware_info)
+  {
+    auto joint_command_interface_descriptions =
+      parse_joint_command_interface_descriptions_from_hardware_info(hardware_info);
+    for (const auto & description : joint_command_interface_descriptions)
+    {
+      joint_command_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
+  /**
+   * Creates all interfaces used for reporting emergency stop, warning and error messages.
+   * The available report interfaces are: EMERGENCY_STOP_SIGNAL, ERROR_SIGNAL, ERROR_SIGNAL_MESSAGE,
+   * WARNING_SIGNAL and WARNING_SIGNAL_MESSAGE. Where the <report_type>_MESSAGE hold the message for
+   * the corresponding report signal.
+   * The interfaces are named like <hardware_name>/<report_interface_type>. E.g. if hardware is
+   * called joint_1 -> interface for WARNING_SIGNAL is called: joint_1/WARNING_SIGNAL
+   */
+  void create_report_interfaces()
+  {
+    // EMERGENCY STOP
+    InterfaceInfo emergency_interface_info;
+    emergency_interface_info.name = hardware_interface::EMERGENCY_STOP_SIGNAL;
+    emergency_interface_info.data_type = "bool";
+    InterfaceDescription emergency_interface_descr(info_.name, emergency_interface_info);
+    emergency_stop_ = std::make_shared<StateInterface>(emergency_interface_descr);
+
+    // ERROR
+    // create error signal interface
+    InterfaceInfo error_interface_info;
+    error_interface_info.name = hardware_interface::ERROR_SIGNAL_INTERFACE_NAME;
+    error_interface_info.data_type = "std::array<uint8_t>";
```

Comment:
When we define those types, we should think about how to name them in URDF too. This looks a bit complicated.

What do you think if we say that the syntax for vectors/arrays is  `vector<_type_>[size]` for example here would be `vector<uint8_t>[32]`. As long as we are pre-initializing the whole memory, by using `reserve` for example`, we should not be breaking real-time capabilities of the system.

```suggestion
    error_interface_info.data_type = "vector<uint8_t>[32]";
```

Reply:
The naming is to complicated. I will change it. But since the size should not be changeable, I would rather go with array then vectors.

Reply:
wouldn't array also be better for realtime because of memory allocation?